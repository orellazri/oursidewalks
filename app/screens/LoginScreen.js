import { useState } from "react";
import { View, SafeAreaView, StyleSheet, Keyboard, ActivityIndicator } from "react-native";
import Toast from "react-native-root-toast";
import { signInWithEmailAndPassword } from "firebase/auth";
import { setItemAsync } from "expo-secure-store";

import { auth } from "../utils/firebase";
import { colors } from "../utils/data";
import ContinueButton from "../components/ContinueButton";
import BackButton from "../components/BackButton";
import TextInput from "../components/TextInput";
import Title from "../components/Title";
import { CustomException, validEmail } from "../utils/utils";
import { useReport } from "../utils/ReportContext";

export default function LoginScreen({ navigation }) {
  const { setUid, retrieveUserInfo } = useReport();

  const [loading, setLoading] = useState(false);

  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  const handleSubmit = async () => {
    try {
      Keyboard.dismiss();

      // Form validation
      if (email == "" || password == "") throw new CustomException("חובה להזין את כל הפרטים");
      if (!validEmail(email)) throw new CustomException("כתובת האימייל שהזנת אינה תקינה");
      if (password.length < 6) throw new CustomException("הסיסמה חייבת להכיל לפחות 6 תווים");

      setLoading(true);

      // Log in
      const credentials = await signInWithEmailAndPassword(auth, email.toLowerCase(), password);

      // Set user in persistent storage and state
      setUid(credentials["user"]["uid"]);
      await setItemAsync("uid", credentials["user"]["uid"]);
      retrieveUserInfo(credentials["user"]["uid"]);

      navigation.navigate("Welcome");
    } catch (e) {
      let errorMessage = "התתחברות נכשלה";
      if (e.code == "auth/wrong-password") errorMessage = "ההתחברות נכשלה. אחד מהפרטים שהזנת לא נכונים";
      if (e instanceof CustomException) errorMessage = e.message;

      Toast.show(errorMessage, {
        duration: Toast.durations.LONG,
        shadow: false,
      });

      setLoading(false);
    }
  };

  if (loading) {
    return (
      <SafeAreaView style={{ flex: 1, justifyContent: "center", alignItems: "center" }}>
        <ActivityIndicator size="large" color="#aaa" />
      </SafeAreaView>
    );
  }

  return (
    <SafeAreaView style={styles.container}>
      {/* Back Button */}
      <BackButton navigation={navigation} />

      {/* Title */}
      <Title text="התחברות" />

      {/* Form */}
      <View style={styles.form}>
        <TextInput label="כתובת אימייל" value={email} onChangeText={setEmail} keyboardType="email-address" autoComplete="email" />
        <TextInput label="סיסמה" value={password} onChangeText={setPassword} secureTextEntry={true} />
      </View>

      {/* Continue Button */}
      <ContinueButton onPress={handleSubmit} />
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    backgroundColor: colors.background,
  },
  form: {
    width: "80%",
    marginTop: "6%",
  },
});
