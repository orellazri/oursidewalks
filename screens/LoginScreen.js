import { useState } from "react";
import { View, Text, SafeAreaView, StyleSheet } from "react-native";

export default function LoginScreen() {
  return (
    <SafeAreaView style={styles.container}>
      <Text>LoginScreen</Text>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    justifyContent: "center",
    alignItems: "center",
  },
});
