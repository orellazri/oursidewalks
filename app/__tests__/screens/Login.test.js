import renderer from "react-test-renderer";
import LoginScreen from "screens/LoginScreen";
import { render } from "../../node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod";

beforeAll(done => {
    done()
})

afterAll(done => {
    done()
})

describe("Login Screen (Negative Testing)",() => {

    it("renders correctly", () => {
        const tree = renderer.create(<LoginScreen />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("fails when all input fields are empty",() => {
        const email = "";
        const password = "";
    });

    it("fails when Email field is empty", () => {
        const email = "";
        const password = "a";
    });

    it("fails when Password field is empty", () => {
        const email = "a";
        const password = "";
    });

    it("fails when Email and Password don't match a registered user", () => {
        const email = "a";
        const password = "a";
    });
})
