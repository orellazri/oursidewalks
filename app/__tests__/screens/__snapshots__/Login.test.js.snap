// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen (Negative Testing) renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F4F4F4",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "marginTop": "10%",
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Assistant-Bold",
          "fontSize": 20,
        }
      }
    >
      התחברות
    </Text>
  </View>
  <View
    style={
      Object {
        "marginTop": "6%",
        "width": "80%",
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "center",
          "marginBottom": 15,
          "width": "80%",
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "flex-start",
            "fontFamily": "Assistant-Bold",
            "fontSize": 17,
            "marginBottom": 5,
          }
        }
      >
        כתובת אימייל
      </Text>
      <TextInput
        allowFontScaling={true}
        autoComplete="email"
        keyboardType="email-address"
        label="כתובת אימייל"
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 10,
            "paddingHorizontal": 10,
            "paddingVertical": 15,
            "textAlign": "right",
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "marginBottom": 15,
          "width": "80%",
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "flex-start",
            "fontFamily": "Assistant-Bold",
            "fontSize": 17,
            "marginBottom": 5,
          }
        }
      >
        סיסמה
      </Text>
      <TextInput
        allowFontScaling={true}
        label="סיסמה"
        onChangeText={[Function]}
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 10,
            "paddingHorizontal": 10,
            "paddingVertical": 15,
            "textAlign": "right",
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "flex-end",
        "backgroundColor": "#FFCB39",
        "borderRadius": 50,
        "flexDirection": "row",
        "marginVertical": "5%",
        "opacity": 1,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Assistant-Bold",
          "fontSize": 18,
          "marginRight": 10,
        }
      }
    >
      המשך
    </Text>
    <Text />
  </View>
</RCTSafeAreaView>
`;
